Ensure new slot is fully processed in `MockPrimaryNode`

NingLin-P opened the pull request:
This PR is part of https://github.com/subspace/subspace/issues/1213

This PR adds support to ensure a new slot is fully processed in `MockPrimaryNode`, this ability along with `wait_for_blocks` in tests can synchronize different events with the testing logic like:
- bundles from the previous slot have been produced and submitted
- receipt-related states like `head_receipt_number`, `state_root`, etc, have updated 
- transactions have been included in the bundle and then applied

which makes the test more deterministic and less flaky.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

liuchengxu commented on test/subspace-test-service/Cargo.toml:
```diff
@@ -16,11 +16,13 @@ targets = ["x86_64-unknown-linux-gnu"]
 futures-timer = "3.0.1"
 rand = "0.8.5"
 pallet-balances = { git = "https://github.com/subspace/substrate", rev = "456cfad45a178617f6886ec400c312f2fea59232" }
+pallet-domains = { version = "0.1.0", default-features = false, path = "../../crates/pallet-domains" }
```
```suggestion
pallet-domains = { version = "0.1.0", path = "../../crates/pallet-domains" }
```

liuchengxu commented on test/subspace-test-service/Cargo.toml:
```diff
@@ -40,6 +42,7 @@ sp-blockchain = { git = "https://github.com/subspace/substrate", rev = "456cfad4
 sp-consensus = { git = "https://github.com/subspace/substrate", rev = "456cfad45a178617f6886ec400c312f2fea59232" }
 sp-consensus-subspace = { version = "0.1.0", path = "../../crates/sp-consensus-subspace" }
 sp-consensus-slots = { version = "0.10.0-dev", default-features = false, git = "https://github.com/subspace/substrate", rev = "456cfad45a178617f6886ec400c312f2fea59232" }
+sp-domains = { version = "0.1.0", default-features = false, path = "../../crates/sp-domains" }
```
```suggestion
sp-domains = { version = "0.1.0", path = "../../crates/sp-domains" }
```

liuchengxu commented on test/subspace-test-service/Cargo.toml:
```diff
@@ -16,11 +16,13 @@ targets = ["x86_64-unknown-linux-gnu"]
 
 [dependencies]
 async-trait = "0.1.58"
+codec = { package = "parity-scale-codec", version = "3.2.1", default-features = false, features = ["derive"] }
```
```suggestion
codec = { package = "parity-scale-codec", version = "3.2.1", features = ["derive"] }
```

liuchengxu commented on test/subspace-test-service/src/mock.rs:
```diff
@@ -178,6 +180,91 @@ impl MockPrimaryNode {
+    }
+
+    /// Wait until the system executor catch up its receipt chain on the primary chain
+    async fn wait_system_executor_catch_up<SClient, SBlock>(
```
Since the goal is to ensure the executor not skip any upcoming slot, `ensure_system_executor_is_fully_synced` sounds more precise and concrete.

liuchengxu commented on test/subspace-test-service/src/mock.rs:
```diff
@@ -178,6 +180,91 @@ impl MockPrimaryNode {
+        }
+
+        let mut import_tx_stream = self.transaction_pool.import_notification_stream();
+        while let Some(ready_tx_hash) = import_tx_stream.next().await {
```
I think this is problematic in theory, the reason is same as the potential missing primary block notifications we had before. When the slot is produced, the executor will be notified instantly and the bundle production starts in the background. If the extrinsic containing the produced bundle is eventually submitted at some point between the last check and this subscription, the notification of that bundle will be lost.

NingLin-P replied:
Make sense! It can also be fixed by moving the subscription in advance, and we don't need to check the extrinsic at the pool anymore.

NingLin-P committed:
Add acknowledgement_sender to the slot notification stream
currently acknowledgement_sender is only intended to use in test to ensure
a slot have fully processed

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Ensure new slot is fully processed in MockPrimaryNode
This commit also come with the is_bundle_present api which is
used in test along with produce_slot to explicitly ensure that
a bundle is produced at the new slot

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Integrate MockPrimaryNode with test invalid_execution_proof_should_not_work

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on crates/subspace-fraud-proof/src/tests.rs:
```diff
@@ -370,12 +366,17 @@ async fn invalid_execution_proof_should_not_work() {
+    futures::join!(
+        alice.wait_for_blocks(1),
+        bob.wait_for_blocks(1),
+        ferdie.produce_n_blocks(1),
```
It could just be named as `ferdie.produce_blocks(1)` for the consistency favor, mostly from a personal preference, non-blocker.

NingLin-P replied:
I have also thought of `ferdie.produce_blocks(1)` but concerned about the confusion between `ferdie.produce_blocks(1)` and `ferdie.produce_block(slot)`.

liuchengxu replied:
Use `produce_block_with_slot(slot)` then.

liuchengxu commented on test/subspace-test-service/src/mock.rs:
```diff
@@ -178,6 +196,30 @@ impl MockPrimaryNode {
     }
+
+    /// Check if a bundle that created by `author_key` at `slot` is present at the transaction pool
+    pub fn is_bundle_present(&self, slot: u64, author_key: Sr25519Keyring) -> bool {
```
Just to be clear.

```suggestion
    pub fn is_bundle_present_in_tx_pool(&self, slot: u64, bundle_author: Sr25519Keyring) -> bool {
```

liuchengxu commented on test/subspace-test-service/src/mock.rs:
```diff
@@ -147,22 +151,36 @@ impl MockPrimaryNode {
-    /// Produce slot
-    pub fn produce_slot(&mut self) -> Slot {
+    /// Produce a slot and wait for the acknowledgement of all the slot notification subscribers
+    pub async fn produce_slot(&mut self) -> Slot {
```
This implies a lot in fact, it won't resolve until the bundle is produced and submitted, let's make it more explicit.

```suggestion
    pub async fn produce_slot_and_wait_for_bundle_submission(&mut self) -> Slot {
```

NingLin-P replied:
Under the context here, `produce_slot` only waits for the acknowledgment of the slot notification subscribers, although we have efforts in other places to ensure bundle submission, it is not clear here, I'm concerned `produce_slot_and_wait_for_bundle_submission` will be a bit confusing, how about `produce_slot_and_wait_for_processing` or `produce_and_process_slot`?

liuchengxu replied:
From the API design's perspective, the consumer of this API honestly doesn't have to know the inner implementation detail IMO. In this scenario, we have a specific need which is to produce a bundle and ensure it's submitted to the primary transaction pool, which is not conveyed in `produce_slot_and_wait_for_processing` and `produce_and_process_slot`. I mean it doesn't have to be generic at present, if it has to be in the future, something like `produce_slot_and_wait_for_slot_handler_completion` could be more explicit to align with our intention.

liuchengxu reviewed (commented):
Some minor comments, otherwise looks okay to me.

NingLin-P committed:
Apply review suggestions

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' into wait-for-bundle-test

NingLin-P committed:
Set primary_block_import_throttling_buffer_size to 0 in test

This will ensure the primary chain will not be ahead of the execution chain by
more than one block, thus slot will not be skipped in test

Signed-off-by: linning <linningde25@gmail.com>
